exquisite_corpus/spm_dataset.py

```diff
@@ -0,0 +1,178 @@
+import numpy as np
+
+from pathlib import Path
+from torch.utils.data import Dataset, BatchSampler, Sampler
+
+
+class SpmIdsDataset(Dataset):
+    """
+    A subclass of Pytorch's Dataset class that exposes a the output of
+    split_sentencepiece_ids and sentencepiece_id_converter.  (Principally,
+    this means a directory full of .npy files containing matrices whose rows
+    are the sequences of SentencePiece id's corresponding to a set of input
+    sentences; we assume each such .npy file corresponds to a distinct length
+    of input sentences.)  Only sentences with sequence length greater than
+    zero are taken to be data items.
+
+    Note:  It is possible to use standard BatchSamplers and RandomSamplers
+    with an SpmIdsDataset, however, the resulting batches may not have the
+    property that all sentences in a batch have the same sequence length.
+    To get uniform sequence lengths in each batch, use an SpmIdsBatchSampler.
+    In particular, if you want to use an SpmIdsDataset with a DataLoader and
+    get uniform sequence length batches, construct the DataLoader with the
+    batch_sampler argument equal to an instance of SpmIdsBatchSampler.
+    """
+
+    def __init__(
+        self,
+        npy_directory,
+        min_length=1,
+        max_length=np.inf,
+        datatype=np.int16,
+        mmap=False,
+    ):
+        self.datatype = datatype
+        self.mmap_mode = "r" if mmap else None
+        npy_root_path = Path(npy_directory)
+        self.npy_paths = {}  # map item lengths to corresponding npy files
+        self.item_counts = {}  # map item lengths to number of corresponding items
+        self.max_item_length = -1  # max length for which there really are any items
```

Comment:
What does it mean for `self.max_item_length` to be -1? That doesn't sound like a maximum length, so is this a flag value or something?

Reply:
It's a value that will be compared with any actual lengths that are seen in the following loop; by setting it less than any possible actual value it is guaranteed to be replaced by any actual value that is encountered.  The alternative, I suppose, would be to set the max length to None here, but then inside the loop the comparison that updates the maximum value would have to have a guard against the possibility that it is (still) None, which makes the code longer and slower.

Reply:
Ah okay, that just needs a comment.

Reply:
Done.